L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -1,110 +1,174 @@
+{ 
 
-    if (triggerTools_.isScintillator(detid)) {
-      HGCScintillatorDetId TC_idV9(detid);
-      return TC_idV9.ietaAbs() << HGCScintillatorDetId::kHGCalPhiOffset | TC_idV9.iphi();  //scintillator
-    } else {
-      HGCalTriggerDetId TC_idV9(detid);
+    for (auto& s: STCs){
 
-      int TC_wafer = TC_idV9.waferU() << kWafer_offset_ | TC_idV9.waferV();
-      int thickness = triggerTools_.thicknessIndex(detid);
+      int thickness = 0;
+      if ( triggerTools_.isSilicon(s.second.getSTCId() ) ){
+	thickness = triggerTools_.thicknessIndex(s.second.getSTCId(),true);
+      }else if ( triggerTools_.isScintillator(s.second.getSTCId()) ){
+	thickness = 3;
+      }
 
-      int TC_12th = 0;
-      int Uprime = 0;
-      int Vprime = 0;
-      int rocnum = detIdToROC_.getROCNumber(TC_idV9.triggerCellU(), TC_idV9.triggerCellV(), 1);
+      std::vector<uint32_t> output_ids = superTCmapping_.getConstituentTriggerCells( s.second.getSTCId() );
+
+      for (const auto& id: output_ids){
+
+	if ( fixedDataSizePerHGCROC_ && thickness > kHighDensityThickness_ && id!=superTCmapping_.getEvenDetId(id)) {
+	  continue;
+	}
+
+        if ( !triggerTools_.validTriggerCell(id) ) {
+          continue;
+        }
+        
+        l1t::HGCalTriggerCell triggerCell;
+        GlobalPoint point = triggerTools_.getTCPosition(id);
+        math::PtEtaPhiMLorentzVector p4(0, point.eta(), point.phi(), 0.);
+        triggerCell.setPosition(point);
+        triggerCell.setP4(p4);
+        triggerCell.setDetId(id);
+        
+        trigCellVecOutput.push_back ( triggerCell );
+        
+	//Change coarse TC positions if needed
+	if ( fixedDataSizePerHGCROC_ == true && thickness > kHighDensityThickness_ ){
+	  coarseTCmapping_.setCoarseTriggerCellPosition( trigCellVecOutput.back() );
+	}
+
+	if ( energyDivisionType_ == oneBitFraction ){ //Get the 1 bit fractions
+
+	  if ( id != s.second.getMaxId() ){
+	    float tc_fraction = getTriggerCellOneBitFraction( s.second.getTCpt(id) , s.second.getSumPt() );
+	    s.second.addToFractionSum(tc_fraction);
+	  }
+	}
 
-      if (rocnum == 1) {
-        Uprime = TC_idV9.triggerCellU();
-        Vprime = TC_idV9.triggerCellV() - TC_idV9.triggerCellU();
+      }
+    }
 
-      } else if (rocnum == 2) {
-        Uprime = TC_idV9.triggerCellU() - TC_idV9.triggerCellV() - 1;
-        Vprime = TC_idV9.triggerCellV();
 
-      } else if (rocnum == 3) {
-        Uprime = TC_idV9.triggerCellU() - kRotate4_;
-        Vprime = TC_idV9.triggerCellV() - kRotate4_;
+    // assign energy
+    for (const  l1t::HGCalTriggerCell & tc : trigCellVecOutput){
+      
+      const auto & stc = STCs[superTCmapping_.getCoarseTriggerCellId(tc.detId())]; 
+    
+      if ( (energyDivisionType_!=superTriggerCell)
+           || ( energyDivisionType_==superTriggerCell && (tc.detId() == stc.getMaxId()) )
+           )  {
+	
+        trigCellVecOutput.push_back( tc );
```

Comment:
It does seem strange, but removing that line seems to have no effect. I'll try to understand why.

Reply:
It had no effect for super TCs, because the earlier TC would never be assigned any energy since the later function acted on `trigCellVecOutput.back()`. However I've changed it now, as it was unclear.

Reply:
Then if just this line is removed I don't understand what happens to other trigger cells for which the `assignSuperTriggerCellEnergy` is not applied.

Reply:
I have now modified this, such that `assignSuperTriggerCellEnergy` is applied for all TCs (and corresponding changes in the function itself.